File: euxis-identity/src/euxis_identity/core/erc8004.py
"""ERC-8004 on-chain agent card generation."""

from __future__ import annotations

from dataclasses import dataclass, field

from euxis_identity.core.did import ServiceEndpoint
from euxis_identity.core.registry import AgentIdentity


@dataclass(frozen=True, slots=True)
class ERC8004AgentCard:
    """ERC-8004 compliant agent identity card."""

    agent_id: str
    agent_uri: str
    metadata: dict[str, object] = field(default_factory=dict)


def generate_agent_card(
    identity: AgentIdentity,
    services: tuple[ServiceEndpoint, ...] = (),
) -> ERC8004AgentCard:
    """Generate an ERC-8004 agent card from a registered identity."""
    metadata: dict[str, object] = {
        "name": identity.metadata.get("name", identity.did),
        "description": identity.metadata.get("description", ""),
        "protocols": ["did:euxis", "erc-8004"],
        "trust_config": {
            "attestation_count": len(identity.attestations),
            "credential_count": len(identity.credentials),
        },
    }

    if services:
        metadata["services"] = [
            {
                "id": s.id,
                "type": s.type,
                "endpoint": s.service_endpoint,
            }
            for s in services
        ]

    return ERC8004AgentCard(
        agent_id=identity.did,
        agent_uri=identity.did,
        metadata=metadata,
    )


def serialize_agent_card(card: ERC8004AgentCard) -> dict[str, object]:
    """Serialize an agent card to a JSON-compatible dict."""
    return {
        "agent_id": card.agent_id,
        "agent_uri": card.agent_uri,
        "metadata": card.metadata,
    }

File: euxis-cpp/euxis-identity-cpp/src/erc8004.cpp
#include "euxis/identity/erc8004.hpp"

#include <format>

#include <spdlog/spdlog.h>

#include "euxis/identity/did.hpp"

namespace euxis::identity {

auto generate_agent_card(
    std::string_view agent_id,
    std::string_view name,
    std::string_view description,
    std::vector<std::string> capabilities,
    std::string_view version
) -> ERC8004AgentCard {
    const auto did = create_did(agent_id);

    ERC8004AgentCard card{
        .agent_id = std::string{agent_id},
        .did = did,
        .name = std::string{name},
        .description = std::string{description},
        .capabilities = std::move(capabilities),
        .version = std::string{version},
        .metadata = nlohmann::json::object(),
    };

    spdlog::debug("Generated ERC-8004 agent card for '{}' (DID: {})", agent_id, did);
    return card;
}

auto ERC8004AgentCard::to_json() const -> nlohmann::json {
    return nlohmann::json{
        {"agentId", agent_id},
        {"did", did},
        {"name", name},
        {"description", description},
        {"capabilities", capabilities},
        {"version", version},
        {"metadata", metadata},
    };
}

} // namespace euxis::identity
